hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/compact/schema/SchemaReplicator.java

```diff
@@ -0,0 +1,364 @@
+    private final Object mutex = new Object();
+    private final ConcurrentHashMap<Long, SchemaReplication> replications = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<Long, CompletableFuture<Void>> inFlightOperations = new ConcurrentHashMap<>();
+
+    // Not final due to late initialization with init method.
+    private NodeEngine nodeEngine;
+
+    public SchemaReplicator(MemberSchemaService schemaService) {
+        this.schemaService = schemaService;
+    }
+
+    /**
+     * Sets the {@link NodeEngine} reference to make the replicator ready to
+     * work.
+     *
+     * @param nodeEngine to set.
+     */
+    public void init(NodeEngine nodeEngine) {
+        this.nodeEngine = nodeEngine;
+    }
+
+    /**
+     * Clears the local state of the replicator.
+     */
+    public void clear() {
+        inFlightOperations.clear();
+        replications.clear();
+    }
+
+    /**
+     * Replicates the schema to the cluster.
+     * <p>
+     * If the schema is known to be already replicated, it returns immediately.
+     * <p>
+     * If not, it initiates the process by:
+     * <ul>
+     *     <li>
+     *         Preparing for the replication process in local by:
+     *         <ul>
+     *             <li>Putting the schema to an in-memory registry</li>
+     *             <li>Persisting the schema to HotRestart(if available and
+     *             enabled)</li>
+     *             <li>Sending the schema to WAN clusters and waiting for it
+     *             to be replicated(if available and enabled)</li>
+     *         </ul>
+     *     </li>
+     *     <li>Marking the replication status of the schema as
+     *     {@link SchemaReplicationStatus#PREPARED}</li>
+     *     <li>Sending the request for preparation to all cluster members.</li>
+     *     <li>On successful acknowledgment from all participants, sending the
+     *     request for marking the schema
+     *     as {@link SchemaReplicationStatus#REPLICATED} to all cluster
+     *     nodes.</li>
+     * </ul>
+     *
+     * @param schema to be replicated.
+     * @return the future which will be completed once the replication process
+     * ends.
+     */
+    public CompletableFuture<Void> replicate(Schema schema) {
+        long schemaId = schema.getSchemaId();
+        if (isSchemaReplicated(schemaId)) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        CompletableFuture<Void> future = inFlightOperations.get(schemaId);
+        if (future != null) {
+            return future;
+        }
+
+        synchronized (mutex) {
+            if (isSchemaReplicated(schemaId)) {
+                return CompletableFuture.completedFuture(null);
+            }
+
+            future = inFlightOperations.get(schemaId);
+            if (future != null) {
+                return future;
+            }
+
+            future = new CompletableFuture<>();
+            inFlightOperations.put(schemaId, future);
+        }
```

Comment:
It is the double-checked locking idiom. Basically, we wanted to avoid taking the lock, if we know that the schema is already replicated or there is an in-flight operation for it